pkg/auth/userregistry/identitymapper/provision.go

```diff
@@ -0,0 +1,153 @@
+	identityregistry "github.com/openshift/origin/pkg/user/registry/identity"
+	userregistry "github.com/openshift/origin/pkg/user/registry/user"
+)
+
+type UserNameGenerator func(base string, sequence int) (string, error)
+
+var (
+	MaxGenerateAttempts = 100
+	DefaultGenerator    = UserNameGenerator(func(base string, sequence int) (string, error) {
+		switch {
+		case sequence == 0:
+			return base, nil
+		case sequence > MaxGenerateAttempts:
+			return "", errors.New("Max attempts exceeded")
+		default:
+			return fmt.Sprintf("%s%d", base, sequence+1), nil
+		}
+	})
+)
+
+type provisioningIdentityMapper struct {
+	identity    identityregistry.Registry
+	user        userregistry.Registry
+	generator   UserNameGenerator
+	initializer user.Initializer
+}
+
+// NewAlwaysCreateUserIdentityToUserMapper creates a mapping for the given identity if it already exists
+func NewAlwaysCreateUserIdentityToUserMapper(identityRegistry identityregistry.Registry, userRegistry userregistry.Registry) authapi.UserIdentityMapper {
+	return &provisioningIdentityMapper{identityRegistry, userRegistry, DefaultGenerator, user.NewDefaultUserInitStrategy()}
+}
+
+// UserFor returns info about the user for whom identity info have been provided
+func (p *provisioningIdentityMapper) UserFor(info authapi.UserIdentityInfo) (kuser.Info, error) {
+	ctx := kapi.NewContext()
+
+	identityName := identityregistry.IdentityName(info.GetProviderName(), info.GetProviderUserName())
+	identity, err := p.identity.GetIdentity(ctx, identityName)
+
+	if kerrs.IsNotFound(err) {
+		return p.createIdentityAndMapping(ctx, info)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return p.getMapping(ctx, identity)
+}
+
+// createIdentityAndMapping creates an identity with a valid user reference for the given identity info
+func (p *provisioningIdentityMapper) createIdentityAndMapping(ctx kapi.Context, info authapi.UserIdentityInfo) (kuser.Info, error) {
+	// Build the part of the identity we know about
+	identity := &userapi.Identity{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: identityregistry.IdentityName(info.GetProviderName(), info.GetProviderUserName()),
+		},
+		ProviderName:     info.GetProviderName(),
+		ProviderUserName: info.GetProviderUserName(),
+		Extra:            info.GetExtra(),
+	}
+
+	// Get or create a persisted user pointing to the identity
+	persistedUser, err := p.getOrCreateUserForIdentity(ctx, identity)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the identity pointing to the persistedUser
+	identity.User = kapi.ObjectReference{
+		Name: persistedUser.Name,
+		UID:  persistedUser.UID,
+	}
+	if _, err := p.identity.CreateIdentity(ctx, identity); err != nil {
+		return nil, err
+	}
+
+	return &kuser.DefaultInfo{
+		Name: persistedUser.Name,
+		UID:  string(persistedUser.UID),
+	}, nil
+}
+
+func (p *provisioningIdentityMapper) getOrCreateUserForIdentity(ctx kapi.Context, identity *userapi.Identity) (*userapi.User, error) {
+
+	for sequence := 0; true; sequence++ {
```

Comment:
Comment here.  Iterate through the max allowed generated user names to see if any of those names are already mapped to our identity.  Names are created in a deterministic order, so the first on that isn't present gets created.  In the case of a race, one will get to persist the user object and the other will fail.